README.md
Graph-Weighted
==============

This is an object oriented, weighted graph implementation inheriting from the
venerable, Perl [Graph](https://metacpan.org/release/Graph) module.


Changes
Revision history for Graph-Weighted:

{{$NEXT}}

0.9101 2020-04-21 01:42:56
    - Use a constant instead of a magic literal in the code.
    - Add the eg/diagram example program.
    - Improve documentation a tiny bit.

0.9100 2020-04-20 22:51:10
    - Add the ->MST_edge_sum method and use it in eg/minimum-spanning-tree.

0.9002 2020-04-20 22:26:10
    - Improve the eg/shortest-path program.
    - Improve documentation ever so slightly.

0.9001 2020-04-20 21:58:08
    - Improve documentation.
    - Improve the eg/minimum-spanning-tree program.
    - Use constant, not Readonly.

0.9   2017-03-07
    - Actually allow non-numeric keys and connections!
    - Display the node label, if there is one, in dump().
    - Add the 02-alpha-keys.t test.

0.8   2017-03-01
    - Add eg/random
    - Add and use the dump() function.

0.7001 2016-02-03
    - Fixed multi-attribute tests.
    - Improved documentation.

0.70  2016-02-02
    - Allowed a graph to have multiple labels.
    - Tested hash attribute/label setting.

0.6001 2016-02-01
    - Improved documentation.
    - Removed the redundant get_weight() method.

0.60  2015-10-14
    - Add support for population with Math::MatrixReal objects.

0.5907 2015-10-09
    - Improved documentation.
    - Updated code in the eg/ examples directory with MST.

0.5906 2015-10-09
    - Improved documentation ever so slightly.

0.5905 2015-10-09
    - Added a handy Shortest Paths example section to the POD.
    - Updated code in the eg/ examples directory with other SP methods.

0.5904 2015-10-05
    - Illustrate SP_Dijkstra in eg/.
    - Improved documentation.

0.5903 2015-10-03
    - Removed the unneeded new() method.

0.5902 2015-10-03
    - Improved documentation.
    - Used Carp::croak instead of warn/die.

0.5901 2015-10-03
    - Removed the old-school "debug mode."
    - Added missing attribute to path cost.

0.59  2015-10-03
    - Added the ability to label nodes.
    - Skip the dist.ini and github readme files.

0.5802 2015-10-02
    - Corrected documentation!

0.5801 2015-10-02
    - Improved documentation.

0.58  2015-10-02
    - Renamed get_attr() to get_cost().
    - Renamed path_attr() to path_cost().

0.57  2015-10-02
    - Renamed span() to vertex_span().
    - Added the edge_span() method.

0.5601 2015-10-02
    - Fixed SYNOPSIS typo.

0.56  2015-10-02
    - Added the path_attr() method.

0.55  2015-09-30
    - Implemented the span() method that returns the lightest and
      heaviest vertices.
    - Vastly simplified the population test.

0.54  2015-09-30
    - Replaced constant with Readonly and base with parent.
    - Improved documentation.
    - Removed the confusing and unnecessary vertex_method and
      edge_method logic.
    - Allow population with Math::Matrix objects.

0.5303 2014-08-30
    - Improved documentation!

0.5301 Aug 8, 2012
    - Improved documentation.
    - Tested for terminal values.

0.53 Aug 8, 2012
    - Added dzil helper plugins to dist creation.
    - Used (and test) both vertex and edge methods!
    - Fixed up POD.
    - Handled terminal nodes for hash datapoints.

0.5201 May 18, 2012
    - Made get_attr() return zero if no attribute value is found.
    - Simplified the example script.
    - Improved POD.

0.52 May 17, 2012
    - Removed polar chart examples.
    - Enabled data hash handling.
    - Improved POD.

0.51 May 15, 2012
    - Allowed weighting of both edges and vertices.

0.5 May 14, 2012
    - Starting over with simplified bits and dzil based distribution.

0.0902_1.tar.gz            29-Sep-2004 19:03  3.2K
0.09.tar.gz                31-Jul-2003 12:18   10K
0.08.tar.gz                11-Jul-2003 16:43   10K
0.07.tar.gz                10-Jul-2003 19:16  9.9K
0.06.tar.gz                10-Jul-2003 09:17  9.0K
0.05.tar.gz                09-Jul-2003 20:38  8.9K
0.4.tar.gz                 05-Jul-2003 09:33   59K
0.3.tar.gz                 04-Jul-2003 19:34   11K
0.2.tar.gz                 04-Jul-2003 16:50   12K
0.01.2.tar.gz              04-Jul-2003 13:16  7.4K
0.01.1.tar.gz              04-Jul-2003 08:39  7.4K
0.01.tar.gz                03-Jul-2003 17:37  7.2K


lib/Graph/Weighted.pm
package Graph::Weighted;

# ABSTRACT: A weighted graph implementation

our $VERSION = '0.9101';

use warnings;
use strict;

use parent qw( Graph );

use Carp qw( croak );

use constant WEIGHT => 'weight';

=head1 SYNOPSIS

 use Graph::Weighted;

 my $gw = Graph::Weighted->new;
 $gw->populate(
    [ [ 0,1,2,0,0 ], # Vertex 0 with 2 edges of weight 3
      [ 1,0,3,0,0 ], #    "   1      2 "               4
      [ 2,3,0,0,0 ], #    "   2      2 "               5
      [ 0,0,1,0,0 ], #    "   3      1 "               1
      [ 0,0,0,0,0 ], #    "   4      0 "               0
    ]
 );
 $gw->dump;

 my ( $lightest, $heaviest ) = $gw->vertex_span;
 ( $lightest, $heaviest ) = $gw->edge_span;

 my $attr = 'probability';
 $gw = Graph::Weighted->new;
 $gw->populate(
    {
        0 => { label => 'A', 1 => 0.4, 3 => 0.6 },
        1 => { label => 'B', 0 => 0.3, 2 => 0.7 },
        2 => { label => 'C', 0 => 0.5, 2 => 0.5 },
        3 => { label => 'D', 0 => 0.2, 1 => 0.8 },
    },
    $attr
 );
 $gw->dump($attr);

 my $cost = $gw->get_cost( [0, 1], $attr );

 $cost = $gw->path_cost( [0, 3, 1, 2], $attr );

 my $tree = $gw->MST_Kruskal;
 my $sum = $gw->MST_edge_sum($tree);

=head1 DESCRIPTION

A C<Graph::Weighted> object is a subclass of the L<Graph> module with attribute
handling.  As such, all of the L<Graph> methods may be used.

=head1 METHODS

=head2 new

  my $gw = Graph::Weighted->new;
  my $gw = Graph::Weighted->new(%arguments);

Return a new C<Graph::Weighted> object.

Please see L<Graph/Constructors> for the possible constructor arguments.

=head2 populate

  $gw->populate($matrix);
  $gw->populate($matrix, $attribute);
  $gw->populate(\@vectors);
  $gw->populate(\@vectors, $attribute);
  $gw->populate(\%data_points);
  $gw->populate(\%data_points, $attribute);

Populate a graph with weighted nodes and edges.

The data can be an arrayref of numeric vectors, a C<Math::Matrix> object, a
C<Math::MatrixReal> object, or a hashref of node-edge values.

Data given as a hash reference may also contain node labels.  Also, the keys
need not be numeric, just unique.

The optional C<attribute> argument is a string with the default C<weight>.

=cut

sub populate {
    my ($self, $data, $attr) = @_;

    # Set the default attribute.
    $attr ||= WEIGHT;

    # What type of data are we given?
    my $data_ref = ref $data;

    if ($data_ref eq 'ARRAY' || $data_ref eq 'Math::Matrix') {
        my $vertex = 0; # Initial vertex id.
        for my $neighbors (@$data) {
            $self->_from_array($vertex, $neighbors, $attr);
            $vertex++; # Move on to the next vertex...
        }
    }
    elsif ($data_ref eq 'Math::MatrixReal') {
        my $vertex = 0;
        for my $neighbors (@{ $data->[0] }) {
            $self->_from_array($vertex, $neighbors, $attr);
            $vertex++;
        }
    }
    elsif ($data_ref eq 'HASH') {
        for my $vertex (keys %$data) {
            for my $entry ( keys %{ $data->{$vertex} } ) {
                if ( $entry eq 'label' ) {
                    my $label = delete $data->{$vertex}{$entry};
                    $self->set_vertex_attribute($vertex, $entry, $label);
                }
            }
            $self->_from_hash( $vertex, $data->{$vertex}, $attr );
        }
    }
    else {
        croak "Unknown data type: $data\n";
    }
}

sub _from_array {
    my ($self, $vertex, $neighbors, $attr) = @_;

    # Initial vertex weight
    my $vertex_weight = 0;

    # Make nodes and edges.
    for my $n (0 .. @$neighbors - 1) {
        my $w = $neighbors->[$n]; # Weight of the edge to the neighbor.
        next unless $w; # Skip zero weight nodes

        # Add a node-node edge to the graph.
        $self->add_edge($vertex, $n);

        $self->set_edge_attribute($vertex, $n, $attr, $w);

        # Tally the weight of the vertex.
        $vertex_weight += $w;
    }

    # Set the weight of the graph node.
    $self->set_vertex_attribute($vertex, $attr, $vertex_weight);
}

sub _from_hash {
    my ($self, $vertex, $neighbors, $attr) = @_;

    # Initial vertex weight
    my $vertex_weight = 0;

    # Handle terminal nodes.
    if (ref $neighbors) {
        # Make nodes and edges.
        for my $n (keys %$neighbors) {
            my $w = $neighbors->{$n}; # Weight of the edge to the neighbor.

            # Add a node-node edge to the graph.
            $self->add_edge($vertex, $n);

            $self->set_edge_attribute($vertex, $n, $attr, $w);

            # Tally the weight of the vertex.
            $vertex_weight += $w;
        }
    }
    else {
        $vertex_weight = $neighbors;
    }

    # Set the weight of the graph node.
    $self->set_vertex_attribute($vertex, $attr, $vertex_weight);
}

=head2 get_cost

  $c = $gw->get_cost($vertex);
  $c = $gw->get_cost($vertex, $attribute);
  $c = $gw->get_cost(\@edge);
  $c = $gw->get_cost(\@edge, $attribute);

Return the named attribute value for the vertex or edge.  If no attribute name
is given, the string C<weight> is used.

=cut

sub get_cost {
    my ($self, $v, $attr) = @_;
    croak 'ERROR: No vertex given to get_cost()' unless defined $v;

    # Default to weight.
    $attr ||= WEIGHT;

    # Return the edge attribute if given a list.
    return $self->get_edge_attribute(@$v, $attr) || 0 if ref $v eq 'ARRAY';

    # Return the vertex attribute if given a scalar.
    return $self->get_vertex_attribute($v, $attr) || 0;
}

=head2 vertex_span

 ($lightest, $heaviest) = $gw->vertex_span;
 ($lightest, $heaviest) = $gw->vertex_span($attr);

Return the lightest and heaviest vertices as array references.

=cut

sub vertex_span {
    my ($self, $attr) = @_;

    # Get the cost of each vertex
    my $mass = {};
    for my $vertex ( $self->vertices ) {
        $mass->{$vertex} = $self->get_cost($vertex, $attr);
    }

    # Find the smallest & biggest costs
    my ($smallest, $biggest);
    for my $vertex ( keys %$mass ) {
        my $current = $mass->{$vertex};
        if ( !defined $smallest || $smallest > $current ) {
            $smallest = $current;
        }
        if ( !defined $biggest || $biggest < $current ) {
            $biggest = $current;
        }
    }

    # Collect the lightest & heaviest vertices
    my ($lightest, $heaviest) = ([], []);
    for my $vertex ( keys %$mass ) {
        push @$lightest, $vertex if $mass->{$vertex} == $smallest;
        push @$heaviest, $vertex if $mass->{$vertex} == $biggest;
    }

    return $lightest, $heaviest;
}

=head2 edge_span

 ($lightest, $heaviest) = $gw->edge_span;
 ($lightest, $heaviest) = $gw->edge_span($attr);

Return the lightest and heaviest edges as array references.

=cut

sub edge_span {
    my ($self, $attr) = @_;

    # Get the cost of each edge
    my $mass = {};
    for my $edge ( $self->edges ) {
        $mass->{ $edge->[0] . '_' . $edge->[1] } = $self->get_cost($edge, $attr);
    }

    # Find the smallest & biggest costs
    my ($smallest, $biggest);
    for my $edge ( keys %$mass ) {
        my $current = $mass->{$edge};
        if ( !defined $smallest || $smallest > $current ) {
            $smallest = $current;
        }
        if ( !defined $biggest || $biggest < $current ) {
            $biggest = $current;
        }
    }

    # Collect the lightest & heaviest edges
    my ($lightest, $heaviest) = ([], []);
    for my $edge ( sort keys %$mass ) {
        my $arrayref = [ split /_/, $edge ];
        push @$lightest, $arrayref if $mass->{$edge} == $smallest;
        push @$heaviest, $arrayref if $mass->{$edge} == $biggest;
    }

    return $lightest, $heaviest;
}


=head2 path_cost

 $c = $gw->path_cost(\@vertices);
 $c = $gw->path_cost(\@vertices, $attr);

Return the summed weight (or cost attribute) of the path edges.

=cut

sub path_cost {
    my ($self, $path, $attr) = @_;

    return undef unless $self->has_path( @$path );

    my $path_cost = 0;

    for my $i ( 0 .. @$path - 2 ) {
        $path_cost += $self->get_cost( [ $path->[$i], $path->[ $i + 1 ] ], $attr );
    }

    return $path_cost;
}

=head2 MST_edge_sum

  $sum = $gw->MST_edge_sum($tree);

Compute the sum of the edges of a minimum-spanning-tree.

=cut

sub MST_edge_sum {
    my ($self, $tree) = @_;

    my $sum = 0;

    my @edges = split /,/, $tree;

    for my $edge (@edges) {
        my @edge = split /=/, $edge;
        $sum += $self->get_cost(\@edge);
    }

    return $sum;
}

=head2 dump

  $gw->dump
  $gw->dump($attr)

Print out the graph showing vertices, edges and costs.

=cut

sub dump {
    my $self = shift;
    my $attr = shift || WEIGHT;

    for my $vertex ( sort { $a <=> $b } $self->vertices ) {
        my $label = $self->get_vertex_attribute($vertex, 'label');
        printf "%svertex: %s %s=%.2f\n",
            ( $label ? "$label " : '' ),
            $vertex,
            $attr,
            $self->get_cost( $vertex, $attr );
        for my $successor ( sort { $a <=> $b } $self->successors($vertex) ) {
            printf "\tedge to: %s %s=%.2f\n",
                $successor,
                $attr,
                $self->get_cost( [ $vertex, $successor ], $attr );
        }
    }
}

1;
__END__

=head1 SEE ALSO

L<Graph>, the parent of this module

L<Graph::Easy::Weighted>, the sibling

The F<eg/*> and F<t/*> programs in this distribution

=cut
